src/sqlfluff/dialects/dialect_snowflake.py

```diff
@@ -7827,3 +7829,38 @@ class AlterMaskingPolicySegment(BaseSegment):
             Sequence("UNSET", "COMMENT"),
         ),
     )
+
+
+class ForInLoopSegment(BaseSegment):
+    """FOR...IN...DO...END FOR statement.
+
+    https://docs.snowflake.com/en/developer-guide/snowflake-scripting/loops#for-loop
+    """
+
+    type = "for_in_statement"
+
+    match_grammar = OneOf(
```

Comment:
Should this not be a Sequence at the top level?

Reply:
It looks like there is a mess with terminators in Delimited() statements and FOR statement itself. Unfortunately, could not find more pretty workaround of that.

Reply:
I think it *should* be possible, but I'll have a go and see if I can make it work that way.

Reply:
Yep - there's something strange going on here. I think it's actually a problem with `ScriptingBlockStatementSegment`. I'm going to have a go at fixing it.

Reply:
Yeah I think there was a bug in the terminator support. I think I've found an fixed it - it also threw up that there's a problem with one of our existing test cases, which I've also solved. @WittierDinosaur I'd love a review given I've added quite a lot of code to this PR.